src/wallet/walletdb.cpp

```diff
@@ -305,8 +305,12 @@ ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue,
             ssKey >> hash;
             CWalletTx wtx;
             ssValue >> wtx;
-            CValidationState state;
-            if (!(CheckRegularTransaction(wtx, state) && (wtx.GetId() == hash) && state.IsValid()))
+            CValidationState state_regular, state_coinbase;
+            if (!(wtx.GetId() == hash))
+                return false;
+            CheckRegularTransaction(wtx, state_regular);
+            CheckCoinbase(wtx, state_coinbase);
+            if (!(state_regular.IsValid() || state_coinbase.IsValid()))
```

Comment:
Why not for for something like

     bool isValid = wtx.IsCoinbase() ? CheckCoinbase(wtx, state) : CheckRegularTransaction(wtx, state);

?